lib/configuration/normalizeConfig.js

```diff
@@ -0,0 +1,184 @@
+  return false;
+}
+
+/**
+ * Appends authorization header when supplied with "user" option.
+ */
+const coerceUserOption = R.when(
+  R.propSatisfies(R.complement(R.isNil), 'user'),
+  R.compose(
+    R.dissoc('user'),
+    R.over(
+      R.lens(updateHeaderWithUser, R.assoc('header')),
+      R.identity
+    )
+  )
+);
+
+const mapIndexed = R.addIndex(R.map);
+
+const coerceApiDescriptions = R.compose(
+  mapIndexed((content, index) => ({
+    location: `configuration.apiDescriptions[${index}]`,
+    content: R.when(R.has('content'), R.prop('content'), content),
+  })),
+  coerceToArray
+);
+
+/**
+ * Coerces the given deprecated value of the "level" option
+ * and returns the supported value for "loglevel" option.
+ */
+const coerceDeprecatedLevelOption = R.when(
+  R.either(R.has('l'), R.has('level')),
+  R.compose(
+    R.dissoc('l'),
+    R.dissoc('level'),
+    R.over(
+      R.lens(
+        R.compose(
+          R.cond([
+            [
+              R.includes(R.__, ['silly', 'debug', 'verbose']),
+              R.always('debug'),
+            ],
+            [R.equals('error'), R.always('error')],
+            [R.equals('silent'), R.always('silent')],
+            [R.T, R.always('warn')],
+          ]),
+          R.either(R.prop('l'), R.prop('level'))
+        ),
+        R.assoc('loglevel')
+      ),
+      R.identity
+    )
+  )
+);
+
+const coerceDataToApiDescriptions = R.compose(
+  R.unnest,
+  R.values,
+  R.evolve({
+    data: R.compose(
+      R.map(([location, content]) => {
+        const nextDesdcription = (typeof content === 'string')
+          ? { location, content }
+          : {
+            location: content.filename,
+            content: content.raw,
+          };
+
+        return nextDesdcription;
+      }),
+      R.toPairs
+    ),
+  }),
+  R.pick(['apiDescriptions', 'data'])
+);
+
+const coerceDeprecatedDataOption = R.when(
+  R.propSatisfies(R.complement(R.isNil), 'data'),
+  R.compose(
+    R.dissoc('data'),
+    R.over(
+      R.lens(
+        coerceDataToApiDescriptions,
+        R.assoc('apiDescriptions')
+      ),
+      R.identity
+    )
+  )
+);
+
+const coerceColorOption = R.when(
+  R.has('c'),
+  R.compose(
+    R.dissoc('c'),
+    R.over(
+      R.lens(R.prop('c'), R.assoc('color')),
+      coerceToBoolean
+    )
+  )
+);
+
+const coerceDeprecatedOptions = R.compose(
+  coerceColorOption,
+  coerceDeprecatedDataOption,
+  coerceDeprecatedLevelOption
+);
+
+const coerceOptions = R.compose(
+  coerceDeprecatedOptions,
+  coerceUserOption,
+  R.evolve({
+    color: coerceToBoolean,
+    apiDescriptions: coerceApiDescriptions,
+    reporter: coerceToArray,
+    output: coerceToArray,
+    header: coerceToArray,
+    method: R.compose(R.map(R.toUpper), coerceToArray),
+    only: coerceToArray,
+    path: coerceToArray,
+    hookfiles: coerceToArray,
+  })
+);
+
+const normalizeConfig = R.compose(
+  coerceOptions,
+  removeUnsupportedOptions
+);
+
+module.exports = {
+  normalizeConfig,
+  _utils: {
```

Comment:
Both ways make them exposed in the module's export. There is a fundamental question about the publicity of this module, as `normalizeConfig` is used internally only.

I am against modifying `normalizeConfig` because it's a *function*, and setting properties on it effectively modifies that function. Named exports keep modules immutable, and serve for access purposes.

Alternatively, we can move all utils into separate directly, and export/import from there. Tests would import utils directly, and main `normalizeConfig` module would import, but won't export them. That's a slight overkill, however, to bring such separation for tests alone.

Reply:
I see your point in having the functions immutable, but I think this isn't really a mutation which would matter, I perceive it more like a workaround, syntax sugar. Bascially, with imports, it's what the "default export" gives you.

My requirements would be developer experience and encapsulation of the code.

I think each module should be encapsulated in a way that it represents one function which is useful outside of that module, is named according to that function, and I can always require it as const fooBarFunction = require('./fooBarFunction.js');` without thinking and without knowing anything about its internals (blackbox with a suitable naming so I can infer what it does without reading the code). And it doesn't matter if that particular function is used just within a certain folder (configuration) or across Dredd (although I think folders should behave in a similar way, hiding things which shouldn't be used by outside code).

If that function composes from helpers relevant only to that function, those should be kept in a file together with it. So far all good. The problem lands when we need to unit test these helper functions (which we definitely should be doing).

I think the only way to retain the blackbox developer experience, the uniform export, and to have the helpers exported and unittested, is to have the "default export" and some hidden ones. While ES6 has a syntax for that, with requires we need a hack: assigning the unittested functions as properties to the main function, which effectively then acts as a default export.

Since we never do this in our code for any other purpose - modifying function properties, I'd say this is an acceptable convention to solve a real problem. Also, while it is mutation of the function instance, I think you could also take it as a way of [reflection](https://en.wikipedia.org/wiki/Reflection_%28computer_programming%29). You are exporting a function, which exposes its helpers in a "protected" property. These helpers belong to that function, they are its parts, so it doesn't seem wrong to me to expose them for the purpose of unit tests in a context of that function.

Reply:
As long as injected util names are not function's native methods (`toString`, `name`, etc.), it should be fine. I will update the pull request.